### DISCUSSION TITLE
next/head <meta> tags only renders Head tags in _app.js but not in other pages

### DISCUSSION THREAD

**Discussion by Chamberlainfrancis** [vercel/next.js#40753] [Help] [answered]
I am creating some sort of micro-blog using `NEXTJS.` I defined some metadata in the _app.js file with overrides for specific components and the article page.

For instance: I have _app.js, home.js and [article].js files, I have added some metadata into the _app.js file as the default metadata and I am overriding it in the [article].js page. 

I have also used the keys tag provided by `NEXTJS `to ensure that the METATAGS do not duplicate themselves. 

When I load the pages in the browser and `inspect` I can see that this works correctly and component-specific overrides are present, but when I `view page source` the changes/overrides on the head tag reverts back to the default settings in the _app.js file.  This affects my SEO and also [Facebook debugger](https://developers.facebook.com/tools/debug) and scrappers because they can only see the default meta tags set on the _app.js page but not the ones specific to a page.

what could be the cause of this, personally I think it's a NEXTJS problem but just writing here to know if there is something I am not doing correctly.

To have a better understanding of what I am talking about I have deployed on vercel for example purposes. please visit this [article page](https://h-blog-cmhu8wnn8-chamberlainfrancis.vercel.app/article/18294/why-you-must-use-a-vpn) on your browser right click and `inspect` the page, on the head section you can see the correct meta tags specific to that page but if you click on `view page source` you can only see the default meta tags set on the _app.js file. 

Thanks for the help in advance.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I did _app like this:

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";
import Head from "next/head";

function MyApp({ Component, pageProps }: AppProps) {
  return (
    <>
      <Head>
        <meta key="foo" property="foo" content="default-value" />
      </Head>
      <Component {...pageProps} />;
    </>
  );
}

export default MyApp;
```


And then two pages, `/`:

```tsx
import Head from "next/head";

const Home = () => {
  return (
    <>
      <Head>
        <meta key="foo" property="foo" content="bar" />
      </Head>
      <div>Home</div>;
    </>
  );
};

export default Home;
```

And a `blog` page:

```tsx
const Blog = () => {
  return <div>Blog</div>;
};

export default Blog;
```

Then I query the page using the REST Client extension in VSCode (that means no browser is ever involved), and I see:

- `/` has the specific meta foo I define, `bar`
- `/blog` has the default meta defined in _app, `default-value`

This is in principle one of the killer features of using a SSR framework like Next.js. From looking at the source code of the example you provide, there's logic in `_app`, which makes it so that the server returns `Loading...` and now the actual page, sort of:

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";
import Head from "next/head";

function MyApp({ Component, pageProps }: AppProps) {
  const isLoading = useSomeHookThatGivesYouTheValue();

  return (
    <>
      <Head>
        <meta key="foo" property="foo" content="default-value" />
      </Head>
      {isLoading ? <div>Loading...</div> : <Component {...pageProps} />};
    </>
  );
}

export default MyApp;
```


```html
<div class="Toastify"></div><div class="container d-flex align-items-center flex-column" style="height:590px"><div class="my-auto text-center"><span class="">Loading...</span></div></div></div>
```

What you need to do is either:

- Detect that the viewing entity is a crawler/bot, and skip the loading screen for them
- Use the `isLoading` condition for the non-head parts of each page, that means, handling it all per page... and let the meta tags always be part of the HTML generated on the server
- Some other solutions I might be forgetting about

## Solution

The problem was that a Context Provider awaited data before rendering its children, which blocked the rendering of the Page Component.

**Reply by Chamberlainfrancis** [reply]
Did you try this ?

```
const Blog = () => {
  return <>
       <Head>
        <meta key="foo" property="foo" content="blog-value" />
      </Head>
      <div>Blog</div>
</>;
};

export default Blog;
```
Also, the loading screen has nothing to do with article.js page here is an actual [example ](https://stackblitz.com/edit/nextjs-5xj6he?file=pages/_app.js) of the code in both files.

PS: Locally, I have removed the loading screen but the results are still the same.

**Reply by icyJoseph** [reply]
I did try that just now and it works too. 

The fact that the application ships a loading screen as starting HTML is the problem, I think. Without a repo to look at, the uglified/minified source points me in that direction. 

It's like, when the server renders the app, for whatever route, it resolves the server HTML to a loading screen, already at the `_app` file level, from what I can see, so anything you do in the page you care about doesn't happen server side. 

It is later on the client that your page contents is rendered. 

## An alternative

It might also be that you have nested `next/head` tags, and I do believe those nullify each other. It might not be obvious how they are nesting though, but essentially if the React component tree resolves to something like:

```tsx
import Head from "next/head";

const Blog = () => {
  return (
    <>
      {/* no bueno, right here /*}
      <Head>
        <Head>
          <meta key="foo" property="foo" content="blog-bar" />
        </Head>
      </Head>
      <div>Home</div>;<div>Blog</div>
    </>
  );
};

export default Blog;
```

**Reply by Chamberlainfrancis** [reply]
I don't know why this is so complicated. However, Locally, I have removed the loading screens from _app.js and tried to follow your examples but the results are still the same.

Also, the information on the meta tags is loaded serverside using `getStaticProps`

**Reply by icyJoseph** [reply]
Would you mind sharing a link to a repo? Or invite me as contributor. Or perhaps you can setup a Stackblitz or GitHub repo with the issue, so we can stop making assumptions. I am out of office now, but I can take a look later.

## Edit

I saw you posted a link to Stackblitz above, I had missed that on my phone. I'll take a look when I am back home.

**Reply by Chamberlainfrancis** [reply]
Okay, I have sent you an invite as a collaborator on the repo. I also set up a branch called local. I look forward to your response, Thanks a lot for the help.

**Reply by icyJoseph** [reply]
Alright, I might be a bit off here, but the context hierarchy of the application contains the `TopicsContextProvider`, which doesn't render its children if `categories` is falsy. Instead it returns a `Loading` component. 

This is why the application is serving a loading screen. I see the request for categories, https://h-blog-cmhu8wnn8-chamberlainfrancis.vercel.app/article/18294/why-you-must-use-a-vpn, resolves and then the meta tags appear.

What do you think?

**Reply by Chamberlainfrancis** [reply]
From my end I have removed `TopicsContextProvider` entirely to confirm if that's the case or the reason, and I can confirm that its not.

I think the problem lies in this piece of code in `pages/article/[...slugs].js`

```
     {router.isFallback ? (
        <span>Loading...</span>
        ) : (
        <>
          <title>{data.title}</title>
          <meta name="title" content={data.title} key="title" />
          <meta name="description" content={data.description} key="desc" />
        </>
      )}
```

I have followed along with this NEXTJS [example blog-starter ](https://github.com/vercel/next.js/blob/canary/examples/blog-starter/pages/posts/%5Bslug%5D.tsx) to match and the issue persists even on the nextJS example

**Reply by icyJoseph** [reply]
If it were `isFallback`, then you don't have to worry about bots because it does `blocking` when bots arrive to the site. 

I think however, that the issue is not that, because when I query as a bot I still see the problem.

I guess it is clear at the moment that the issue is that some component in the hierarchy, when the page is built, at `next build` is resolving to `Loading...` as the actual content of the page, so that whenever we make an HTTP request to it, we see everything else that was rendered around `Loading...`, and even though the `getStaticProps` result is properly inlined as JSON to the page, because the Page Component doesn't get rendered, the props are missing... 

Now the question is where in the tree are we blocking the actual page from flushing out during development. 

In other words, if you inspect `.next/pages/server` folder, after building, you'll find the HTML for articles that just contains `Loading...` - Gonna take a deep dive into the app now, it might just be that there's multiple places where we hang on Loading during build time.
